container/openejb-core/src/main/java/org/apache/openejb/cdi/CdiBuilder.java

```diff
@@ -62,6 +64,7 @@ public void build(final AppInfo appInfo, final AppContext appContext, final List
 
     public static synchronized ThreadSingletonService initializeOWB() {
         logger.info("Created new singletonService " + SINGLETON_SERVICE);
+        SystemInstance.get().addObserver(SINGLETON_SERVICE);
```

Comment:
Why this?

Reply:
ThreadSingletonService will be notified when it is removed during server shutdown and the code added in ThreadSingletonService will remove the associated ThreadContextListener from listeners set. 
See change in ThreadSingletonServiceImpl

Reply:
ok